// @midwayjs/web/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startCluster = exports.Configuration = exports.Agent = exports.Application = exports.createAgentWorkerLoader = exports.createAppWorkerLoader = exports.createEggAgent = exports.createEggApplication = exports.Framework = void 0;
const egg_cluster_1 = require("egg-cluster");
const path_1 = require("path");
__exportStar(require("./interface"), exports);
var web_1 = require("./framework/web");
Object.defineProperty(exports, "Framework", { enumerable: true, get: function () { return web_1.MidwayWebFramework; } });
var base_1 = require("./base");
Object.defineProperty(exports, "createEggApplication", { enumerable: true, get: function () { return base_1.createEggApplication; } });
Object.defineProperty(exports, "createEggAgent", { enumerable: true, get: function () { return base_1.createEggAgent; } });
Object.defineProperty(exports, "createAppWorkerLoader", { enumerable: true, get: function () { return base_1.createAppWorkerLoader; } });
Object.defineProperty(exports, "createAgentWorkerLoader", { enumerable: true, get: function () { return base_1.createAgentWorkerLoader; } });
var application_1 = require("./application");
Object.defineProperty(exports, "Application", { enumerable: true, get: function () { return application_1.Application; } });
Object.defineProperty(exports, "Agent", { enumerable: true, get: function () { return application_1.Agent; } });
var configuration_1 = require("./configuration");
Object.defineProperty(exports, "Configuration", { enumerable: true, get: function () { return configuration_1.EggConfiguration; } });
__exportStar(require("./decorator"), exports);
function startCluster(serverConfig, callback) {
    if (!serverConfig['require']) {
        serverConfig['require'] = [];
    }
    if (!Array.isArray(serverConfig['require'])) {
        serverConfig['require'] = [serverConfig['require']];
    }
    serverConfig['require'].push((0, path_1.join)(__dirname, 'cluster'));
    return (0, egg_cluster_1.startCluster)(serverConfig, callback);
}
exports.startCluster = startCluster;
