// utility/dist/commonjs/crypto.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.base64decode = exports.base64encode = exports.hmac = exports.sha512 = exports.sha256 = exports.sha1 = exports.md5 = exports.hash = void 0;
const node_crypto_1 = require("node:crypto");
const node_crypto_2 = __importDefault(require("node:crypto"));
const nativeHash = 'hash' in node_crypto_2.default ? node_crypto_2.default.hash : null;
/**
 * hash
 *
 * @param {String} method hash method, e.g.: 'md5', 'sha1'
 * @param {String|Buffer|ArrayBuffer|TypedArray|DataView|Object} s input value
 * @param {String} [format] output string format, could be 'hex' or 'base64'. default is 'hex'.
 * @return {String} md5 hash string
 * @public
 */
function hash(method, s, format) {
    if (s instanceof ArrayBuffer) {
        s = Buffer.from(s);
    }
    const isBuffer = Buffer.isBuffer(s) || ArrayBuffer.isView(s);
    if (!isBuffer && typeof s === 'object') {
        s = JSON.stringify(sortObject(s));
    }
    if (nativeHash) {
        // try to use crypto.hash first
        // https://nodejs.org/en/blog/release/v21.7.0#crypto-implement-cryptohash
        return nativeHash(method, s, format);
    }
    const sum = (0, node_crypto_1.createHash)(method);
    sum.update(s, isBuffer ? 'binary' : 'utf8');
    return sum.digest(format || 'hex');
}
exports.hash = hash;
/**
 * md5 hash
 *
 * @param {String|Buffer|Object} s input value
 * @param {String} [format] output string format, could be 'hex' or 'base64'. default is 'hex'.
 * @return {String} md5 hash string
 * @public
 */
function md5(s, format) {
    return hash('md5', s, format);
}
exports.md5 = md5;
/**
 * sha1 hash
 *
 * @param {String|Buffer|Object} s input value
 * @param {String} [format] output string format, could be 'hex' or 'base64'. default is 'hex'.
 * @return {String} sha1 hash string
 * @public
 */
function sha1(s, format) {
    return hash('sha1', s, format);
}
exports.sha1 = sha1;
/**
 * sha256 hash
 *
 * @param {String|Buffer|Object} s input value
 * @param {String} [format] output string format, could be 'hex' or 'base64'. default is 'hex'.
 * @return {String} sha256 hash string
 * @public
 */
function sha256(s, format) {
    return hash('sha256', s, format);
}
exports.sha256 = sha256;
/**
 * sha512 hash
 *
 * @param {String|Buffer|Object} s input value
 * @param {String} [format] output string format, could be 'hex' or 'base64'. default is 'hex'.
 * @return {String} sha512 hash string
 * @public
 */
function sha512(s, format) {
    return hash('sha512', s, format);
}
exports.sha512 = sha512;
/**
 * HMAC algorithm.
 *
 * Equal bash:
 *
 * ```bash
 * $ echo -n "$data" | openssl dgst -binary -$algorithm -hmac "$key" | openssl $encoding
 * ```
 *
 * @param {String} algorithm dependent on the available algorithms supported by the version of OpenSSL on the platform.
 *   Examples are 'sha1', 'md5', 'sha256', 'sha512', etc.
 *   On recent releases, `openssl list-message-digest-algorithms` will display the available digest algorithms.
 * @param {String} key the hmac key to be used.
 * @param {String|Buffer} data content string.
 * @param {String} [encoding='base64'] default encoding is base64
 * @return {String} digest string.
 */
function hmac(algorithm, key, data, encoding) {
    encoding = encoding || 'base64';
    const hmac = (0, node_crypto_1.createHmac)(algorithm, key);
    hmac.update(data, Buffer.isBuffer(data) ? 'binary' : 'utf8');
    return hmac.digest(encoding);
}
exports.hmac = hmac;
/**
 * Base64 encode string.
 *
 * @param {String|Buffer} s input value
 * @param {Boolean} [urlSafe=false] Encode string s using a URL-safe alphabet,
 *   which substitutes - instead of + and _ instead of / in the standard Base64 alphabet.
 * @return {String} base64 encode format string.
 */
function base64encode(s, urlSafe) {
    if (!Buffer.isBuffer(s)) {
        s = Buffer.from(s);
    }
    let encode = s.toString('base64');
    if (urlSafe) {
        encode = encode.replace(/\+/g, '-').replace(/\//g, '_');
    }
    return encode;
}
exports.base64encode = base64encode;
/**
 * Base64 string decode.
 *
 * @param {String} encodeStr base64 encoding string.
 * @param {Boolean} [urlSafe=false] Decode string s using a URL-safe alphabet,
 *   which substitutes - instead of + and _ instead of / in the standard Base64 alphabet.
 * @param {encoding} [encoding=utf8] if encoding = buffer, will return Buffer instance
 * @return {String|Buffer} plain text.
 */
function base64decode(encodeStr, urlSafe, encoding) {
    if (urlSafe) {
        encodeStr = encodeStr.replace(/\-/g, '+').replace(/_/g, '/');
    }
    const buf = Buffer.from(encodeStr, 'base64');
    if (encoding === 'buffer') {
        return buf;
    }
    return buf.toString(encoding || 'utf8');
}
exports.base64decode = base64decode;
function sortObject(o) {
    if (!o || Array.isArray(o) || typeof o !== 'object') {
        return o;
    }
    const keys = Object.keys(o);
    keys.sort();
    const values = [];
    for (const k of keys) {
        values.push([k, sortObject(o[k])]);
    }
    return values;
}
