// utility/dist/commonjs/number.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.random = exports.toSafeNumber = exports.isSafeNumberString = exports.MAX_SAFE_INTEGER_STR = exports.MIN_SAFE_INTEGER = exports.MAX_SAFE_INTEGER = void 0;
// http://www.2ality.com/2013/10/safe-integers.html
// http://es6.ruanyifeng.com/#docs/number
exports.MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || Math.pow(2, 53) - 1;
exports.MIN_SAFE_INTEGER = -exports.MAX_SAFE_INTEGER;
exports.MAX_SAFE_INTEGER_STR = String(exports.MAX_SAFE_INTEGER);
const MAX_SAFE_INTEGER_STR_LENGTH = exports.MAX_SAFE_INTEGER_STR.length;
/**
 * Detect a number string can safe convert to Javascript Number.
 *
 * @param {String} s number format string, like `"123"`, `"-1000123123123123123123"`
 */
function isSafeNumberString(s) {
    if (s[0] === '-') {
        s = s.substring(1);
    }
    if (s.length < MAX_SAFE_INTEGER_STR_LENGTH ||
        (s.length === MAX_SAFE_INTEGER_STR_LENGTH && s <= exports.MAX_SAFE_INTEGER_STR)) {
        return true;
    }
    return false;
}
exports.isSafeNumberString = isSafeNumberString;
/**
 * Convert string to Number if string in safe Number scope.
 *
 * @param {String} s number format string.
 * @return {Number|String} success will return Number, otherwise return the original string.
 */
function toSafeNumber(s) {
    if (typeof s === 'number') {
        return s;
    }
    return isSafeNumberString(s) ? Number(s) : s;
}
exports.toSafeNumber = toSafeNumber;
/**
 * Produces a random integer between the inclusive `lower` and `upper` bounds.
 *
 * @param {Number} lower The lower bound.
 * @param {Number} upper The upper bound.
 * @return {Number} Returns the random number.
 */
function random(lower, upper) {
    // random()
    if (lower === undefined) {
        return 0;
    }
    // random(lower) => random(0, lower)
    if (upper === undefined) {
        upper = lower;
        lower = 0;
    }
    let temp;
    // random(upper, lower)
    if (lower > upper) {
        temp = lower;
        lower = upper;
        upper = temp;
    }
    return Math.floor(lower + Math.random() * (upper - lower));
}
exports.random = random;
